bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,350 @@
+                // Note: IOUtils.copyBytes with length == 0 will enter an infinite loop, hence the need for this check.
+                return 0;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException e) {
+            throw convertException(handle.getChunkName(), "doWrite", e);
+        }
+        return length;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException {
+        ensureInitializedAndNotClosed();
+        int length = 0;
+        try {
+            val sources = new Path[chunks.length - 1];
+            this.fileSystem.truncate(getFilePath(chunks[0].getName()), chunks[0].getLength());
+            for (int i = 1; i < chunks.length; i++) {
+                val chunkLength = chunks[i].getLength();
+                this.fileSystem.truncate(getFilePath(chunks[i].getName()), chunkLength);
+                sources[i - 1] = getFilePath(chunks[i].getName());
+                length += chunkLength;
+            }
+            // Concat source file into target.
+            this.fileSystem.concat(getFilePath(chunks[0].getName()), sources);
+        } catch (IOException e) {
+            throw convertException(chunks[0].getName(), "doConcat", e);
+        }
+        return length;
+    }
+
+    @Override
+    protected boolean doTruncate(ChunkHandle handle, long offset) throws UnsupportedOperationException {
+        throw new UnsupportedOperationException(getClass().getName() + " does not support chunk truncation.");
+    }
+
+    @Override
+    protected void doSetReadOnly(ChunkHandle handle, boolean isReadOnly) throws ChunkStorageException {
+        try {
+            this.fileSystem.setPermission(getFilePath(handle.getChunkName()), isReadOnly ? READONLY_PERMISSION : READWRITE_PERMISSION);
+        } catch (IOException e) {
+            throw convertException(handle.getChunkName(), "doSetReadOnly", e);
+        }
+    }
+
+    //endregion
+
+    //region Storage Implementation
+
+    @SneakyThrows(IOException.class)
+    public void initialize() {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkState(this.fileSystem == null, "HDFSStorage has already been initialized.");
+        Configuration conf = new Configuration();
+        conf.set("fs.default.name", this.config.getHdfsHostURL());
+        conf.set("fs.default.fs", this.config.getHdfsHostURL());
+        conf.set("fs.hdfs.impl", "org.apache.hadoop.hdfs.DistributedFileSystem");
```

Comment:
what about org.apache.hadoop.hdfs.DistributedFileSystem.class.getName() ? 
this way we will have a compile time guarantee that the name is good

Reply:
This sounds sensible. 

@sachin-j-joshi since this this not new code (it's borrowed from the other HDFSStorage.java), may I suggest collecting all these HDFS improvement ideas (there's at least one other below) in a new GitHub issue and we can address them later? 

This particular change is unrelated to PDP34 so I don't want to tackle it here.

Reply:
Fixed
conf.set("fs.hdfs.impl", org.apache.hadoop.hdfs.DistributedFileSystem.class.getName());